File: 백준/Gold/10868. 최솟값/최솟값.py
from sys import stdin
from math import log2, ceil
input = stdin.readline

def init(node,st,ed):
    if st == ed:
        tree_min[node] = num[st]
        return tree_min[node] 
    mid = (st + ed) // 2
    tree_min[node] = min(init(node*2,st,mid),init(node*2+1,mid+1,ed))
    return tree_min[node]   

def min_query(node,st,ed,l,r):
    if st > r or ed < l:
        return 1000000001
    if l <= st and ed <= r:
        return tree_min[node]
    mid = (st + ed) // 2
    return min(min_query(node*2,st,mid,l,r),min_query(node*2+1,mid+1,ed,l,r))
    
n,m = map(int,input().split())
h = int(ceil(log2(n)))
tree_size = 1 << (h+1) # 2의 (h+1) 제곱, 비트 시프트
num = []
tree_min = [0] * tree_size

for _ in range(n):
    num.append(int(input()))

init(1,0,n-1)

for _ in range(m):
    a,b = map(int,input().split())
    print(min_query(1,0,n-1,a-1,b-1))

File: 백준/Gold/10868. 최솟값/최솟값.java
import java.io.*;
import java.util.*;

public class Main {
	static int N;
	static long tree[];

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		StringBuilder sb = new StringBuilder();
		N = Integer.parseInt(st.nextToken());
		int M = Integer.parseInt(st.nextToken());
		int h = (int) Math.ceil(Math.log(N) / Math.log(2));
		int treeSize = 1 << (h + 1);
		int startIdx = treeSize / 2 - 1;
		tree = new long[treeSize + 1];
		Arrays.fill(tree, Integer.MAX_VALUE);
		for (int i = startIdx + 1; i <= startIdx + N; i++) {
			tree[i] = Long.parseLong(br.readLine());
		}
		init(treeSize);
		for (int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			sb.append(findMin(a + startIdx, b + startIdx)).append("\n");
		}
		System.out.print(sb);
	}

	static void init(int i) {
		while (i != 1) {
			if (tree[i / 2] > tree[i])
				tree[i / 2] = tree[i];
			i--;
		}
	}

	static long findMin(int s, int e) {
		long minValue = Integer.MAX_VALUE;
		while (s <= e) {
			if (s % 2 == 1) {
				minValue = Math.min(minValue, tree[s]);
				s++;
			}
			if (e % 2 == 0) {
				minValue = Math.min(minValue, tree[e]);
				e--;
			}
			s = s / 2;
			e = e / 2;
		}
		return minValue;
	}

}